lib/als/async-hook.js

```diff
@@ -0,0 +1,150 @@
+import asyncHooks from 'async_hooks';
+import { getActiveEvent, getInfo } from './als';
+import { EventType } from '../constants';
+import { Ntp } from '../ntp';
+import { checkForNested } from '../tracer/tracer';
+import { stackTrace } from '../utils';
+
+const getResource = (asyncId) => {
+  if (!asyncId) return;
+
+  const info = getInfo();
+
+  if (!info) return;
+
+  return info.resources.get(asyncId);
+};
+
+export const isDescendantOf = (asyncId, parentAsyncId) => {
+  const resource = getResource(asyncId);
+
+  if (!resource) return false;
+
+  if (resource.triggerAsyncId === parentAsyncId) return true;
+
+  const parent = getResource(resource.triggerAsyncId);
+
+  if (!parent) return false;
+
+  return isDescendantOf(parent.asyncId, parentAsyncId);
+};
+
+export const getAncestryChainOfIds = (asyncId) => {
+  const resource = getResource(asyncId);
+
+  if (!resource) return [];
+
+  if (!resource.triggerAsyncId) return [asyncId];
+
+  const parent = getResource(resource.triggerAsyncId);
+
+  if (!parent) return [asyncId];
+
+  return [...getAncestryChainOfIds(parent.asyncId), asyncId];
+};
+
+export const getActiveEventResource = () => {
+  const activeEvent = getActiveEvent();
+
+  return getResource(activeEvent?.asyncId);
+};
+
+const captureStartTime = (asyncId) => {
+  const resource = getResource(asyncId);
+
+  if (!resource) return;
+
+  resource.startTime = Ntp._now();
+};
+
+const captureEndTime = (asyncId) => {
+  const resource = getResource(asyncId);
+  if (!resource) return;
+
+  // Some resources do not trigger before/after hooks.
+  if (!resource.startTime) {
```

Comment:
This is interesting. Do you have any idea on when the hooks are or are not triggered?

Reply:
No, it is in my list for further investigation

Reply:
I have a much better picture now, they are not triggered when the promise resolves to a primitive or a value that is not another promise.